lib/serialization/v1_adapter.rb

```diff
@@ -109,21 +109,33 @@ def attributes_for_serializer(serializer, options)
     end
 
     def include_assoc?(assoc)
-      return false unless @options[:include]
-      check_assoc("#{assoc}$")
+      return false unless @options.key?(:include)
+
+      check_assoc(assoc)
     end
 
     def include_nested_assoc?(assoc)
-      return false unless @options[:include]
-      check_assoc("#{assoc}.")
+      return false unless @options.key?(:include)
+
+      check_nested_assoc(assoc)
     end
 
     def check_assoc(assoc)
       include_opt = @options[:include]
       include_opt = include_opt.split(',') if include_opt.is_a?(String)
-      include_opt.any? do |s|
-        s.match(/^#{assoc.gsub('.', '\.')}/)
-      end
+
+      # check if any of the included associations match the current association
+      include_opt.any?(assoc)
+    end
+
+    def check_nested_assoc(assoc)
+      include_opt = @options[:include]
+      include_opt = include_opt.split(',') if include_opt.is_a?(String)
+
+      # check if any of the included associations match the prefix path of the association
+      # i.e. ${association}.path
+      nested_assoc_pattern = "#{assoc}."
+      include_opt.any? { |included_assoc| included_assoc.starts_with?(nested_assoc_pattern) }
```

Comment:
I'm curious to know in what cases this is ever true. (i.e.
 ``` 
nested_assoc_pattern = "#{assoc}."
include_opt.any? { |included_assoc| included_assoc.starts_with?(nested_assoc_pattern)
```

Reply:
it's not that i could find - this was what i was referencing when i said
> The code in this class is largely unused and can be refactored down even further (i've done this and might PR it after this lands), however it's largely untested but we do have high level tests on output formats so that's useful.

I have another branch where i remove the unused code and refactor but I didn't have any trust in it as we don't have any tests on this code...so I left it with this refactor for the security issue fix. 

I can push that code up as a draft PR if you fancy looking at it and we can move forward if you think it's worthwhile?